### PR TITLE
Linux Auto-Updater

### PR DESCRIPTION
**Pull request by compsup**
## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# Proposed Changes
- Linux auto-updater

## Checklist:

- [X] My code follows the style guidelines of this project and have read CONTRIBUTING.md
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules
- [X] I have checked my code and corrected any misspellings

## Example of what it looks like
image

### CODE REVIEW

**Comment by Kurdakov** [comment]
why do you copy only main file? in fact many files might change in a new version.

**Comment by compsup** [comment]
Because the only files left are the config files, and if anything gets updated (Ex: Config files) the program needs to do the conversion on first startup which is not my problem. And if a new file(s) is needed it should be created on first startup also.

**Review by TechStudent11 — CHANGES_REQUESTED** [review]
Isn't there a way in the GitHub API to pull the latest release?

**Comment by compsup** [comment]
> Isn't there a way in the GitHub API to pull the latest release?

That's currently what is does. The updater fetches the latest tag and then inputs that into GitHubs API too pull down the latest .tar.gz file

**Comment by compsup** [comment]
I should point out, currently the updater assumes that there is only going to be one main file, which currently is true, but if the main files gets split into smaller ones the script will have to be updated (most likely to copy all python files over using a loop).

**Comment by TechStudent11** [comment]
> > Isn't there a way in the GitHub API to pull the latest release?
> 
> That's currently what is does. The updater fetches the latest tag and then inputs that into GitHubs API too pull down the latest .tar.gz file

Oh okay. I only read the first commit so I didn't assume you fixed that.

> I should point out, currently the updater assumes that there is only going to be one main file, which currently is true, but if the main files gets split into smaller ones the script will have to be updated (most likely to copy all python files over using a loop).

That's true and that is something that is planned (well, I think so).

**Comment by compsup** [comment]
I know I have heard some talk about the main file being split. But considering it would be a big update (backwards compatibility breaking) I think it will be a bit before we have to worry about that, and when it does happen the updater can be revised.

**Comment by compsup** [comment]
Looking at v2.10 the .tar.gz file when unzipped does not have an extra folder which will break the updater, before this gets merged there needs to be a standard for these updates. @ThioJoe

**Comment by TechStudent11** [comment]
@compsup Is this really just for Linux? Wouldn't it work for Windows or Mac?

**Comment by compsup** [comment]
> @compsup Is this really just for Linux? Wouldn't it work for Windows or Mac?

It would most likely work on mac os because its *nix (if someone wants to test it, it can be a mac os installer also). In Windows the .exe is much simpler but in theory it could work however I cannot test it. The other issue with having it work for Windows is that you would have to choose between the .exe or raw installation which makes things much more confusing.

**Comment by compsup** [comment]
> @compsup that's true. But you could use the sys module to detect if it's an EXE or `.py`.

My point is that Thio is maintaining the Windows updater while I am maintaining the Linux/Mac os updater. I do not want to make this more complicated then it needs to be. Windows updater uses the .exe and linux/mac os uses the raw .tar.gz file.

**Comment by TechStudent11** [comment]
> > @compsup that's true. But you could use the sys module to detect if it's an EXE or `.py`.
> 
> My point is that Thio is maintaining the Windows updater while I am maintaining the Linux/Mac os updater. I do not want to make this more complicated then it needs to be. Windows updater uses the .exe and linux/mac os uses the raw .tar.gz file.

Ah, okay. I got it.

**Comment by Firecul** [comment]
> I just mean for the end user on Windows they don't care how it installs they just care how they run it and generally a .EXE file is much easier to run for the end-user.

I'd have to agree, in Manjaro I'm using the .py and everything installs and runs easily. In Windows I'm just using the exe because it wasn't working easily and I couldn't be bothered fighting the ease of use aspect.

**Comment by compsup** [comment]
Fixed merge conflicts

**Comment by ThioJoe** [comment]
> I think it will be a bit before we have to worry about that, and when it does happen the updater can be revised.

bruh moment https://github.com/ThioJoe/YT-Spammer-Purge/releases/tag/v2.11.0-Beta2

Also the assets folder cannot be ignored, as it holds the default config and stuff which are occasionally updated. And probably the requirements.txt in case that changes.

**Comment by TechStudent11** [comment]
> > I think it will be a bit before we have to worry about that, and when it does happen the updater can be revised.
> 
> bruh moment https://github.com/ThioJoe/YT-Spammer-Purge/releases/tag/v2.11.0-Beta2

bruh

**Comment by compsup** [comment]
@ThioJoe Would you be alright to move all config files over to a config folder? Like move secrets, token, and config into a `./config` folder. This would simplify updating by a lot because we can just copy everything over except the `./config`.

**Comment by compsup** [comment]
This fixes all past problems, however it requires a code refactor with all config files being put into `./config`. I cannot test if this patch actually works until that change is made. When that change is made I will make sure that my code works.

**Comment by ryleu** [comment]
This is not a good way to manage this. A pypi package would be a much better solution. Linux "auto updates" tend to break on systems that aren't POSIX compliant, while a pypi package and a version check on each run would work just fine.

**Review by berkctezc — APPROVED** [review]

**Comment by oittaa** [comment]
> This is not a good way to manage this. A pypi package would be a much better solution. Linux "auto updates" tend to break on systems that aren't POSIX compliant, while a pypi package and a version check on each run would work just fine.

The whole `install_full.sh` looks overly complicated and brittle way to install anything. Is there some reason to _not_ distribute as a pypi package like @ryleu said?

**Comment by ryleu** [comment]
> Is there some reason to _not_ distribute as a pypi package like @ryleu said?

I’ll go through what I think some arguments could be:

1. API token management would be a pain if it wasn’t in a user-determined directory.

Solution: use environment variables or arguments. An environment variable is the preferred method for running code in containers, which is where this package is most likely to run. If you are opposed to these, for whatever reason, Linux has a set directory where things go: `~/.config`.

2. It’d be hard.

It’s not. I know a 12 year old who did it.

3. This method already works.

Sure, but it’s not ideal. It can’t be set up to update with the rest of the system and it makes it more difficult to install a specific version of the package if that’s needed for whatever reason. This way of doing things also wouldn’t work in, for example, a Heroku worker dyno.

**Comment by compsup** [comment]
Here are my 2 cents. I have to agree that the current way of installing the package on linux is brittle and the auto updater I created here is not the proper way of doing things on linux. While a pypi package is fine I propose the program just gets packaged in distros package managers, like AUR for arch and PPA's on debian. That way the program gets updated with the system and is therefor much less likely to break.

**Comment by jyooru** [comment]
> While a pypi package is fine I propose the program just gets packaged in distros package managers, like AUR for arch and PPA's on debian.

That's not really a reason not to push to PyPI.

The work done to package for PyPI is often used when packaging for distributions. For example:

- [pytest's package on Arch](https://github.com/archlinux/svntogit-community/blob/413f83a1a6afa61d4bd2c98e4e9e956c4877ac72/trunk/PKGBUILD#L37) uses `setup.py` to build:
    ```sh
    build() {
      cd pytest-$pkgver
      python setup.py build
    }
    ```
- [pytest's package on Nix](https://github.com/NixOS/nixpkgs/blob/27249995942ba1718f7f05075d44c93bb8e0267c/pkgs/development/python-modules/pytest/default.nix#L26) downloads from PyPI:
    ```nix
    buildPythonPackage rec {
      pname = "pytest";
      version = "6.2.5";
      disabled = !isPy3k;

      src = fetchPypi {
        inherit pname version;
        sha256 = "131b36680866a76e6781d13f101efb86cf674ebb9762eb70d3082b6f29889e89";
      };
    ```

**Comment by ryleu** [comment]
> …I propose the program just gets packaged in distros package managers, like AUR for arch and PPA's on debian. 

That would work, but there are a lot of distros, so it would be much more work to maintain. PPAs are also a bad idea because they’re a massive security issue. Coprs on DNF distros are similar.

>That way the program gets updated with the system and is therefor much less likely to break.

This is the detriment of pip. It’s terrible at dependency management.

**Comment by rachmadaniHaryono** [comment]
> This is the detriment of pip. It’s terrible at dependency management.

if this program can be installed from pypi, i recommend to add documentation to install the program with pipx

this way the program dependency can be isolated

also for development i create #459 so developer can easily use it with poetry

**Review by EthanHindmarsh — APPROVED** [review]
LGTM! ✅

Though, are checks for failed updates needed? Or, would a failed update cause an error?

**Comment by TechStudent11** [comment]
That's a funny tag

image

I don't think there is any harm in merging. Maybe into a beta version just to see how it works out on others machines.

**Comment by compsup** [comment]
This PR has turned into a debate of the best way too update a program on linux instead of giving feedback on the code. I'm not saying that this debate is invalid just could we maybe either accept or reject this PR and then move on too deciding the best way of going about this. Maybe create a new issue regarding updating on linux.

Regarding the pypi packaging. That sounds like a great way. If someone who knows how to do that does that, that would be great, otherwise I will in the next upcoming days take a look at that.

**Comment by TechStudent11** [comment]
> This PR has turned into a debate of the best way too update a program on linux instead of giving feedback on the code. I'm not saying that this debate is invalid just could we maybe either accept or reject this PR and then move on too deciding the best way of going about this. Maybe create a new issue regarding updating on linux.

I don't even understand half of what this debate is about.

> Regarding the pypi packaging. That sounds like a great way. If someone who knows how to do that does that, that would be great, otherwise I will in the next upcoming days take a look at that.

It's pretty simple to create at PyPI package from my understanding.

File: Scripts/files.py [hunk 1]
```diff
@@ -296,10 +299,71 @@ def check_for_update(currentVersion, updateReleaseChannel, silentCheck=False):
...
+          # Name of this file
+          # Temp folder for update
+          stagingFolder = "temp"
+
+          # Fetch the latest update
+          print(f"\n> Downloading version: {F.GREEN}{latestVersion}{S.R}")
+
+          url = f'https://codeload.github.com/ThioJoe/YT-Spammer-Purge/tar.gz/refs/tags/v{latestVersion}'
+          r = requests.get(url, stream=True)
+          if(r.status_code == 200):
```
Comment (Dobby233Liu) [inline Scripts/files.py:316]:
```suggestion
          if r.status_code == 200:
```
The code above don't have parentheses, removing them should be fine

File: Scripts/files.py [hunk 2]
```diff
@@ -296,10 +299,71 @@ def check_for_update(currentVersion, updateReleaseChannel, silentCheck=False):
...
+          # Extract the tar file and delete it
+          print("\n> Extracting...")
+          with tarfile.open(tarFileName) as file:
+            file.extractall(f'./{stagingFolder}')
+          os.remove(tarFileName)
+          print(f"> Installing...")
+          # Retrieve the name of the folder containing the main file, we are assuming there will always be only one folder here
+          extraFolderPath = os.listdir(f"./{stagingFolder}")
+          # If there happens to be more then one folder
+          if(len(extraFolderPath) != 1):
```
Comment (Dobby233Liu) [inline Scripts/files.py:336]:
```suggestion
          if len(extraFolderPath) != 1:
```

**Review by Dobby233Liu — COMMENTED** [review]
I have some minor formatting nitpick, otherwise this looks good to me

**Comment by KendallDoesCoding** [comment]
> That's a funny label
> 
> image
> 
> I don't think there is any harm in merging. Maybe into a beta version just to see how it works out on others machines.

According to me, I think @ThioJoe added that label because he uses Windows, so he wants people that use Linux to let him know if it works properly.

**Comment by dada513** [comment]
This is a good feature IMO
It doesn't matter that distributing apps on Linux the right way handles updates.
Most pepole have it installed via Python right now, and might not want to use native packaging for a simple script
PyPI would be nice though definietly, since its just a `pip3 install` away

Sure, using native methods like AUR and COPR is better, but still most users will have it installed via python.
I could make an AUR, makedeb (better than PPA, see https://makedeb.org) and COPR packages. I would need to be able to disable the updater though in those packages while building.

**Comment by ThioJoe** [comment]
Alright I'll merge it into beta branch then

**Pull request merged**